packages/fields/src/actions/set-as-homepage.tsx

```diff
@@ -0,0 +1,192 @@
+			useDispatch( coreStore );
+		const { createSuccessNotice, createErrorNotice } =
+			useDispatch( noticesStore );
+
+		async function onSetPageAsHomepage( event: React.FormEvent ) {
+			event.preventDefault();
+
+			try {
+				// If selected page is set to draft, publish the page.
+				if ( item.status === 'draft' ) {
+					await saveEntityRecord( 'postType', 'page', {
+						...item,
+						status: 'publish',
+					} );
+				}
+
+				// Save new home page settings.
+				await saveEntityRecord( 'root', 'site', {
+					page_on_front: item.id,
+					show_on_front: 'page',
+				} );
+
+				closeModal?.();
+
+				await saveEditedEntityRecord( 'root', 'site', undefined, {
+					page_on_front: item.id,
+					show_on_front: 'page',
+				} );
+
+				createSuccessNotice( __( 'Homepage updated' ), {
+					type: 'snackbar',
+				} );
+
+				onActionPerformed?.( items );
+			} catch ( error ) {
+				const typedError = error as CoreDataError;
+				const errorMessage =
+					typedError.message && typedError.code !== 'unknown_error'
+						? typedError.message
+						: __( 'An error occurred while setting the homepage' );
+				createErrorNotice( errorMessage, { type: 'snackbar' } );
+			}
+		}
+
+		const renderModalBody = () => {
+			if ( 'posts' === showOnFront ) {
+				return (
+					<>
+						<Text>
+							{ sprintf(
+								// translators: %s: title of the page to be set as the homepage.
+								__(
+									'Set "%s" as the site homepage? This will replace the current homepage which is set to display latest posts.'
+								),
+								pageTitle
+							) }
+						</Text>
+					</>
+				);
+			}
+
+			const modalTranslatedString =
+				// translators: %1$s: title of page to be set as the home page. %2$s: title of the current home page.
+				__(
+					'Set "%1$s" as the site homepage? This will replace the current homepage: "%2$s"'
+				);
+
+			return (
+				<Text>
+					{ sprintf(
+						modalTranslatedString,
+						pageTitle,
+						currentHomePageTitle
+					) }{ ' ' }
+				</Text>
+			);
+		};
+
+		return (
+			<form onSubmit={ onSetPageAsHomepage }>
+				<VStack spacing="5">
+					{ renderModalBody() }
+					<HStack justify="right">
+						<Button
+							__next40pxDefaultSize
+							variant="tertiary"
+							onClick={ () => {
+								closeModal?.();
+							} }
+						>
+							{ __( 'Cancel' ) }
+						</Button>
+						<Button
+							__next40pxDefaultSize
+							variant="primary"
+							type="submit"
+						>
+							{
+								// translators: Button to confirm setting the specified page as the homepage.
+								__( 'Set homepage' )
+							}
+						</Button>
+					</HStack>
+				</VStack>
+			</form>
+		);
+	};
+
+const setAsHomepage: Action< PostWithPermissionsAndContext > = {
+	id: 'set-as-homepage',
+	label: __( 'Set as homepage' ),
+	isEligible( post ) {
+		const { pageOnFront, pageForPosts } =
+			post.additionalContext?.siteSettings;
+		const { hasFrontPageTemplate } = post.additionalContext?.themeInfo;
```

Comment:
Just a little detail and definitely not a strong opinion, but can we just add all the data we need as keys within the `additionalContext`? It seems a bit unnecessary to group them by sources (where they came from: `siteSettings` or `themeInfo`).